staging/src/k8s.io/apimachinery/pkg/util/net/http.go

```diff
@@ -132,13 +133,31 @@ func SetTransportDefaults(t *http.Transport) *http.Transport {
 	if s := os.Getenv("DISABLE_HTTP2"); len(s) > 0 {
 		klog.Infof("HTTP2 has been explicitly disabled")
 	} else if allowsHTTP2(t) {
-		if err := http2.ConfigureTransport(t); err != nil {
+		if err := configureHTTP2Transport(t); err != nil {
 			klog.Warningf("Transport failed http2 configuration: %v", err)
 		}
 	}
 	return t
 }
 
+func configureHTTP2Transport(t *http.Transport) error {
+	t2, err := http2.ConfigureTransports(t)
+	if err != nil {
+		return err
+	}
+	// The following enables the HTTP/2 connection health check added in
+	// https://github.com/golang/net/pull/55. The health check detects and
+	// closes broken transport layer connections. Without the health check,
+	// a broken connection can linger too long, e.g., a broken TCP
+	// connection will be closed by the Linux kernel after 13 to 30 minutes
+	// by default, which caused
+	// https://github.com/kubernetes/client-go/issues/374 and
+	// https://github.com/kubernetes/kubernetes/issues/87615.
+	t2.ReadIdleTimeout = 30 * time.Second
```

Comment:
I'm assuming you validated this works properly with watches. I wonder if we will end up wanting these parameters to be tunable. Also a little concerned we have no way to turn this off if we have problems with it.

Reply:
I manually verified that `kubectl get -w` worked fine after there was no traffic for >30s.

Hopefully the default values of these parameters work for all clusters. I added environment variables to allow tuning these parameters or even disable the feature.